### PR TITLE
Implemented Ghost Palace exchange NPC

### PR DESCRIPTION
**Pull request by Atemo**
* **Addressed Issue(s)**: #3936

* **Server Mode**: renewal

* **Description of Pull Request**: 

- [x] Requires https://github.com/rathena/rathena/pull/6625

Thanks to @zackdreaver and https://www.divine-pride.net/ !

### CODE REVIEW

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	mes "Adventurer friend <ITEM>[Gray Shard]<INFO>6672</INFO></ITEM> is what I need. How about exchange it with something I have?";
+	next;
+	.@type = select( "Show me your weapon.", "Show me your armor.", "Anything special?", "I do not need." ) - 1;
+	mes "[A traveler of dimension]";
+	switch(.@type) {
+	case 0:
+		mes "Oh! Are you interested in weapons? We have all the weapons we can get from the Ghost Palace. Why don't you choose one?";
+		break;
+	case 1:
+		mes "Armor? We have all the weapons we can get from the Ghost Palace. Why don't you choose one?";
```
Comment (zackdreaver) [inline npc/re/merchants/ghost_palace_exchange.txt]: ``We have all the armors``

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		case 1:
+			break;
+		case 2:
+			mes "[A traveler of dimension]";
+			switch(.@type) {
+			case 0:
+				mes "I have more than enough weapons.";
+				break;
+			case 1:
+				mes "We have many armors for you. Why don't you pic one?";
```
Comment (zackdreaver) [inline npc/re/merchants/ghost_palace_exchange.txt]: ``pick``

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+				break;
+			case 2:
+				mes "If you bring me Gray Shard, I will any you anything that you want.";
```
Comment (zackdreaver) [inline npc/re/merchants/ghost_palace_exchange.txt]: ``I will any you anything that you want`` to ``I will trade it with anything you want`` ?

**Comment by zackdreaver** [comment]
IIRC kRO changed gray shard weight to 0.1.
Maybe you can also update it?

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+//= Walkthrough Conversion.
+//  Note: Gray Charcoal items are disabled
+//		(currently not implemented).
+//===== Changelogs: ==========================================
+//= 1.0 First Version. [Capuche]
+//============================================================
+
+dali02,51,130,4	script	Dimension Traveler	4_F_SHABBY,{
+	mes "[Dimension Traveler]";
+	mes "Adventurer friend <ITEM>[Gray Shard]<INFO>6672</INFO></ITEM> is what I need. How about exchange it with something I have?";
```
Comment (aleos89) [inline npc/re/merchants/ghost_palace_exchange.txt]:
```suggestion
	mes "Adventurer friend, <ITEM>[Gray Shard]<INFO>6672</INFO></ITEM> is what I need. How about exchanging it with something I have?";
```

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		mes "Oh! Are you interested in weapons? We have all the weapons we can get from the Ghost Palace. Why don't you choose one?";
+		break;
+	case 1:
+		mes "Armor? We have all the armors we can get from the Ghost Palace. Why don't you choose one?";
+		break;
+	case 2:
+		mes "Huhh! It seems like many of them already heard that I have the best stuff. Feel free to check out my collections.";
+		break;
+	case 3:
+		mes "Good by friend~ Come again!";
```
Comment (aleos89) [inline npc/re/merchants/ghost_palace_exchange.txt]:
```suggestion
		mes "Goodbye friend~ Come again!";
```

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+		close;
+	}
+	next;
+	mes "[Dimension Traveler]";
+	mes "Ah! The number that is written beside will be needed";
```
Comment (aleos89) [inline npc/re/merchants/ghost_palace_exchange.txt]:
```suggestion
	mes "Ah! The number that is written beside <ITEM>[Gray Shard]<INFO>6672</INFO></ITEM> will be needed.";
```

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	mes "<ITEM>[Gray Shard]<INFO>6672</INFO></ITEM> Please keep that in mind.";
```
Comment (aleos89) [inline npc/re/merchants/ghost_palace_exchange.txt]:
```suggestion
	mes "Please keep that in mind.";
```

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	for ( .@i = 0; .@i < .@size; .@i += 2 )
+		.@menu$ += getitemname(.@item[.@i]) + " - " + .@item[.@i+1] + ":";
+
+	while(true) {
+		next;
+		.@s = (select(.@menu$) - 1) * 2;
+		mes "[Dimension Traveler]";
+		if (.@type != 1)	// armor type doesn't display item info
+			mes "Do you mean <ITEM>[" + getitemname(.@item[.@s]) + "]<INFO>" + .@item[.@s] + "</INFO></ITEM>?";
+		mes "" + .@item[.@s+1] + " ^006400Gray Shard^000000 is required for exchange.";
```
Comment (aleos89) [inline npc/re/merchants/ghost_palace_exchange.txt]:
```suggestion
		mes "" + .@item[.@s+1] + " ^006400Gray Shard^000000 are required in exchange.";
```

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+		next;
+		switch( select( "Let's exchange.", "Let me see other stuffs.", "I will come back later" ) ) {
```
Comment (aleos89) [inline npc/re/merchants/ghost_palace_exchange.txt]:
```suggestion
		switch( select( "Let's exchange.", "Let me see other stuff.", "I will come back later." ) ) {
```

File: npc/re/merchants/ghost_palace_exchange.txt [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+			}
+			continue;
+		case 3:
+			mes "[Dimension Traveler]";
+			mes "Take care friend~ See you next time!";
+			close;
+		}
+		if (countitem(6672) < .@item[.@s+1]) {
+			mes "[Dimension Traveler]";
+			mes "You put me in difficult situation, friend.";
```
Comment (aleos89) [inline npc/re/merchants/ghost_palace_exchange.txt]:
```suggestion
			mes "You put me in a difficult situation, friend.";
```

**Comment by Angelic234** [comment]
.@item should be .@items

**Comment by Angelic234** [comment]
All of .@item should be .@items

**Comment by Questune09** [comment]
I think there's another one npc missing on this.or this is a custom one?
source from : https://www.novaragnarok.com/wiki/Ghost_Palace
Credits to owner.

image
image
image

**Comment by Atemo** [comment]
The Dimension Traveler npc only gives Grey Abrasive to enchant the tanos weapons.
I don't have the info on how the weapon is enchanted however the Grey Abrasive are type consumable, I guess a menu appears when the player consumes them.

Additional info for the interested : https://www.divine-pride.net/forum/index.php?/topic/3378-kro-geffen-magic-tournament-and-ghost-palace-enchant/

**Comment by RagnaWay** [comment]
Add this items on item_db
https://www.divine-pride.net/database/item/23675/
https://www.divine-pride.net/database/item/23676/
https://www.divine-pride.net/database/item/23677/
https://www.divine-pride.net/database/item/23678/
https://www.divine-pride.net/database/item/23679/

**Comment by RagnaWay** [comment]
Update!

**Pull request merged**